### PR TITLE
Don't scope decimals as illegal; tokenize decimal marks, illegal identifiers

### LINKED ISSUE #318: Inconsistent scoping of decimals
I noticed some bugs on decimal numbers:

c

```javascript
0.9 == .9

9.0 == 9.
```

a

b

On the first line, `.9` shouldn't be illegal: [Floating-point literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Grammar_and_types#Floating-point_literals).

Also, periods in numbers should have their own scope IMO, different from `constant.numeric`, the same way plus and minus signs are `keyword.operator`.

### PR DESCRIPTION
**Pull request by MaximSokolov**
```js
// octal
015, 0010, 0o77, 0O10

// hexadecimal
0xFF, 0xff, 0x1123, 0X00111, 0XF1A7

// binary
0b11, 0b0011, 0b11,                 

// decimal
// [(+|-)][digits][.digits][(E|e)[(+|-)]digits]
1E+12
3.1415926
.123456789
3.1E+12
.1e-23

1
1.
.1
1.1
1E+3
1.E3
.1E+3
1.1E+3

// illegal identifiers
1abc
123abc
1$
123$abc
.1a
obj.1
obj.1a
1$.123
```

Fixes #318

### CODE REVIEW

File: spec/javascript-spec.coffee [hunk 1]
```diff
@@ -1220,11 +1224,6 @@ describe "Javascript grammar", ->
       expect(tokens[3]).toEqual value: '.', scopes: ['source.js', 'meta.delimiter.property.period.js']
       expect(tokens[4]).toEqual value: 'b', scopes: ['source.js', 'variable.other.property.js']
 
-      {tokens} = grammar.tokenizeLine('a.123illegal')
```
Comment (50Wliu) [inline spec/javascript-spec.coffee]: Is this impossible to test for now?  Can't we just give numbers higher priority?

File: grammars/javascript.cson [hunk 2]
```diff
@@ -918,8 +918,35 @@
...
+              (?:\\b[0-9]+)
+            )
+            ([eE][+-]?[0-9]+)?
+          )
+        '''
+        'captures':
+          '0':
+            'name': 'constant.numeric.decimal.js'
+          '1':
+            'name': 'meta.delimiter.decimal-mark.period.js'
```
Comment (50Wliu) [inline grammars/javascript.cson]: I'm not sold on `decimal-mark` - it sounds really weird to me.  Maybe just `decimal`.

**Comment by MaximSokolov** [comment]
updated

**Comment by 50Wliu** [comment]
Awesome!  @chbk would you be so kind to test this as well? :stuck_out_tongue:

File: grammars/javascript.cson [hunk 3]
```diff
@@ -918,8 +922,42 @@
...
+        'name': 'constant.numeric.binary.js'
+      }
+      {
+        'match': '\\b(?<!\\$)0(o|O)?[0-7]+\\b(?!\\$)'
+        'name': 'constant.numeric.octal.js'
+      }
+      {
+        'match': '''(?x)
+          (?<!\\$)(?:
+            (?:\\b[0-9]+(\\.)[0-9]+[eE][+-]?[0-9]+\\b)| # 1.1E+3
```
Comment (50Wliu) [inline grammars/javascript.cson]: `1.E3` is also possible...it doesn't look like you're catching that right now

**Comment by chbk** [comment]
Tested and working :+1:

**Comment by 50Wliu** [comment]
Thanks for testing @chbk!

**Pull request merged**